"""
Simple gRPC Client Ä‘á»ƒ test server
"""
import grpc
import sys
import os

# ThÃªm project root vÃ o Python path
project_root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.insert(0, project_root)

from src.proto import kvstore_pb2
from src.proto import kvstore_pb2_grpc


class KVStoreClient:
    """Client Ä‘á»ƒ connect Ä‘áº¿n server vÃ  gá»i RPC methods"""
    
    def __init__(self, host='localhost', port=8001):
        """
        Khá»Ÿi táº¡o connection Ä‘áº¿n server
        
        Args:
            host: Server hostname (default: localhost)
            port: Server port (default: 8001)
        """
        # Táº¡o channel (connection) Ä‘áº¿n server
        self.address = f'{host}:{port}'
        self.channel = grpc.insecure_channel(self.address)
        
        # Táº¡o stub (Ä‘áº¡i diá»‡n cho remote service)
        self.stub = kvstore_pb2_grpc.KeyValueStoreStub(self.channel)
        
        print(f"âœ… Connected to server at {self.address}")
    
    def put(self, key, value):
        """Gá»­i PUT request Ä‘áº¿n server"""
        print(f"ğŸ“¤ Sending PUT: {key} = {value}")
        request = kvstore_pb2.PutRequest(key=key, value=value)
        response = self.stub.Put(request)
        print(f"âœ… PUT successful")
        return response
    
    def get(self, key):
        """Gá»­i GET request Ä‘áº¿n server"""
        print(f"ğŸ“¥ Sending GET: {key}")
        request = kvstore_pb2.GetRequest(key=key)
        response = self.stub.Get(request)
        print(f"âœ… GET successful")
        return response
    
    def delete(self, key):
        """Gá»­i DELETE request Ä‘áº¿n server"""
        print(f"ğŸ—‘ï¸ Sending DELETE: {key}")
        request = kvstore_pb2.DeleteRequest(key=key)
        response = self.stub.Delete(request)
        print(f"âœ… DELETE successful")
        return response
    
    def list_keys(self):
        """Gá»­i ListKeys request Ä‘áº¿n server"""
        print(f"ğŸ“‹ Sending ListKeys")
        request = kvstore_pb2.ListKeysRequest()
        response = self.stub.ListKeys(request)
        print(f"âœ… ListKeys successful")
        return response
    
    def close(self):
        """ÄÃ³ng connection"""
        self.channel.close()
        print("ğŸ”Œ Connection closed")


# Test script khi cháº¡y file nÃ y trá»±c tiáº¿p
if __name__ == '__main__':
    # Táº¡o client instance
    client = KVStoreClient()
    
    print("\nğŸ§ª Testing basic operations...\n")
    
    # Test PUT
    client.put("user:1", "Alice")
    
    # Test GET
    client.get("user:1")
    
    # Test DELETE
    client.delete("user:1")
    
    # Test ListKeys
    client.list_keys()
    
    # Close connection
    client.close()
    
    print("\nâœ… All tests completed!")
